diff -ruaN rr264x-linux-src-v1.5.orig/inc/linux/Makefile.def rr264x-linux-src-v1.5/inc/linux/Makefile.def
--- rr264x-linux-src-v1.5.orig/inc/linux/Makefile.def	2012-07-19 10:47:34.000000000 +0200
+++ rr264x-linux-src-v1.5/inc/linux/Makefile.def	2013-09-29 19:08:31.819426622 +0200
@@ -74,8 +74,15 @@
 KERNELDIR := /lib/modules/$(shell uname -r)/build
 endif
 
-MAJOR := $(shell expr `grep LINUX_VERSION_CODE $(KERNELDIR)/include/linux/version.h | cut -d\  -f3` / 65536 % 65536)
-MINOR := $(shell expr `grep LINUX_VERSION_CODE $(KERNELDIR)/include/linux/version.h | cut -d\  -f3` / 256 % 256)
+VERSION_H := $(shell sh -c " \
+if test -f $(KERNELDIR)/include/linux/version.h; then \
+    echo $(KERNELDIR)/include/linux/version.h; \
+elif test -f $(KERNELDIR)/include/generated/uapi/linux/version.h; then \
+    echo $(KERNELDIR)/include/generated/uapi/linux/version.h; \
+fi")
+
+MAJOR := $(shell expr `grep LINUX_VERSION_CODE $(VERSION_H) | cut -d\  -f3` / 65536 % 65536)
+MINOR := $(shell expr `grep LINUX_VERSION_CODE $(VERSION_H) | cut -d\  -f3` / 256 % 256)
 KERNEL_VER := $(MAJOR).$(MINOR)
 
 ifeq ($(KERNEL_VER),)
@@ -153,8 +160,8 @@
 # export variables used by the scripts.
 export HPT_ROOT TARGETNAME TARGETMODS TARGETOBJS KERNELDIR C_DEFINE KERNEL_VER
 
-#patchkernel:
-#	@sh $(HPT_ROOT)/osm/linux/patch.sh
+patchkernel:
+	@sh $(HPT_ROOT)/osm/linux/patch.sh
 
 install: $(TARGET)
 	@sh $(HPT_ROOT)/osm/linux/install.sh
diff -ruaN rr264x-linux-src-v1.5.orig/osm/linux/hptinfo.c rr264x-linux-src-v1.5/osm/linux/hptinfo.c
--- rr264x-linux-src-v1.5.orig/osm/linux/hptinfo.c	2012-07-19 10:47:35.000000000 +0200
+++ rr264x-linux-src-v1.5/osm/linux/hptinfo.c	2013-09-29 19:02:56.949128290 +0200
@@ -8,6 +8,7 @@
 #include "osm_linux.h"
 #include "hptintf.h"
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 typedef struct _hpt_GET_INFO {
 	char *buffer;
 	int buflength;
@@ -52,6 +53,10 @@
 	hpt_copy_mem_info(pinfo, buf, len);
 	return len;
 }
+#else // >= KERNEL_VERSION(3,10,0)
+typedef struct seq_file HPT_GET_INFO;
+#define hpt_copy_info seq_printf
+#endif
 
 #define HPT_DO_IOCTL(code, inbuf, insize, outbuf, outsize) ({\
 	IOCTL_ARG arg;\
@@ -181,7 +186,7 @@
 					devinfo.u.device.ControllerId+1,
 					devinfo.u.device.PathId+1,
 					devinfo.u.device.TargetId+1,
-					devinfo.u.device.IdentifyData.ModelNumber
+					(char*)devinfo.u.device.IdentifyData.ModelNumber
 				);
 		else {
 			memcpy(sn, devinfo.u.device.IdentifyData.SerialNumber,
@@ -193,7 +198,7 @@
 				(devinfo.u.device.Flags & DEVICE_FLAG_IN_ENCLOSURE) ? "E" : "",
 				devinfo.u.device.PathId+1,
 				devinfo.u.device.TargetId+1,
-				devinfo.u.device.IdentifyData.ModelNumber, sn,
+				(char*)devinfo.u.device.IdentifyData.ModelNumber, sn,
 				(int)(devinfo.Capacity*512/1000000),
 				(devinfo.u.device.Flags & DEVICE_FLAG_DISABLED)? "Disabled" : "Normal",
 				devinfo.u.device.ReadAheadEnabled? "[RA]":"",
@@ -234,15 +239,25 @@
 }
 
 #define MAX_PHYSICAL_DEVICE	128
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 int hpt_proc_get_info(struct Scsi_Host *host, char *buffer, char **start, off_t offset, int length)
+#else
+int hpt_proc_show_info(struct seq_file *m, struct Scsi_Host *host)
+#endif
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 	HPT_GET_INFO info;
+#else
+#define info (*m)
+#define buffillen count ? 0 : 0
+#endif
 	int i, j, count;
 	CONTROLLER_INFO_V2 conInfo2;
 	DEVICEID *ids;
 	int devs;	
 	PVBUS_EXT vbus_ext = get_vbus_ext(host);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 	info.buffer     = buffer;
 	info.buflength  = length;
 	info.bufoffset  = offset;
@@ -250,6 +265,7 @@
 	info.buffillen  = 0;
 
 	if (start) *start = buffer;
+#endif
 
 	hpt_copy_info(&info, "%s %s\n\n", driver_name_long, driver_ver);
 
diff -ruaN rr264x-linux-src-v1.5.orig/osm/linux/os_linux.c rr264x-linux-src-v1.5/osm/linux/os_linux.c
--- rr264x-linux-src-v1.5.orig/osm/linux/os_linux.c	2012-07-19 10:47:35.000000000 +0200
+++ rr264x-linux-src-v1.5/osm/linux/os_linux.c	2013-09-29 19:04:46.305899005 +0200
@@ -203,7 +203,11 @@
 
 	if (page)
 		return (PageHighMem(page)?
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0)
+				(char *)kmap_atomic(page) :
+#else
 				(char *)kmap_atomic(page, HPT_KMAP_TYPE) :
+#endif
 				(char *)page_address(page))
 			+ (psg->addr.bus & 0xffffffff);
 	else
@@ -213,7 +217,11 @@
 void os_kunmap_sgptr(void *ptr)
 {
 	if ((HPT_UPTR)ptr >= (HPT_UPTR)high_memory)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0)
+		kunmap_atomic(ptr);
+#else
 		kunmap_atomic(ptr, HPT_KMAP_TYPE);
+#endif
 }
 #else 
 void *os_kmap_sgptr(PSG psg) { return psg->addr._logical; }
diff -ruaN rr264x-linux-src-v1.5.orig/osm/linux/osm_linux.c rr264x-linux-src-v1.5/osm/linux/osm_linux.c
--- rr264x-linux-src-v1.5.orig/osm/linux/osm_linux.c	2012-07-19 10:47:35.000000000 +0200
+++ rr264x-linux-src-v1.5/osm/linux/osm_linux.c	2013-09-29 19:07:42.957528927 +0200
@@ -449,7 +449,11 @@
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)
 	struct scatterlist *sg;
 	sg = scsi_sglist(cmd);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0)
+	*pbuf = kmap_atomic(HPT_SG_PAGE(sg)) + sg->offset;
+#else
 	*pbuf = kmap_atomic(HPT_SG_PAGE(sg), HPT_KMAP_TYPE) + sg->offset;
+#endif
 	buflen = sg->length;
 #else 
 
@@ -479,7 +483,11 @@
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)
 	struct scatterlist *sg;
 	sg = scsi_sglist(cmd);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0)
+	kunmap_atomic((char *)buf - sg->offset);
+#else
 	kunmap_atomic((char *)buf - sg->offset, HPT_KMAP_TYPE);
+#endif
 #else 
 
 	if (cmd->use_sg) {
@@ -1799,6 +1807,7 @@
 	return -EINVAL;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 static int hpt_proc_info26(struct Scsi_Host *host, char *buffer, char **start,
 					off_t offset, int length, int inout)
 {
@@ -1808,6 +1817,7 @@
 	else
 		return hpt_proc_get_info(host, buffer, start, offset, length);
 }
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
 static int hpt_proc_info24(char *buffer,char **start, off_t offset,
@@ -2108,7 +2118,12 @@
 	#endif
 #else /* 2.6.x */
 	proc_name:               driver_name,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 	proc_info:               hpt_proc_info26,
+#else
+	show_info:               hpt_proc_show_info,
+	write_info:              hpt_proc_set_info,
+#endif
 	max_sectors:             128,
 #endif
 	this_id:                 -1
diff -ruaN rr264x-linux-src-v1.5.orig/osm/linux/osm_linux.h rr264x-linux-src-v1.5/osm/linux/osm_linux.h
--- rr264x-linux-src-v1.5.orig/osm/linux/osm_linux.h	2012-07-19 10:47:35.000000000 +0200
+++ rr264x-linux-src-v1.5/osm/linux/osm_linux.h	2013-09-29 19:09:28.890307159 +0200
@@ -175,7 +175,9 @@
 #define scsi_set_max_cmd_len(host, len) host->max_cmd_len = len
 #endif
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0)
+#define HPT_FIND_PCI_DEVICE pci_get_device
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,0)
 #define HPT_KMAP_TYPE KM_BIO_SRC_IRQ
 #define HPT_FIND_PCI_DEVICE pci_get_device
 #else 
@@ -297,6 +299,9 @@
 void refresh_sd_flags(PVBUS_EXT vbus_ext);
 void hpt_do_ioctl(IOCTL_ARG *ioctl_args);
 void hpt_stop_tasks(PVBUS_EXT vbus_ext);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0)
 int hpt_proc_get_info(struct Scsi_Host *host, char *buffer, char **start, off_t offset, int length);
-
+#else
+int hpt_proc_show_info(struct seq_file *m, struct Scsi_Host *host);
+#endif
 #endif
diff -ruaN rr264x-linux-src-v1.5.orig/osm/linux/patch.sh rr264x-linux-src-v1.5/osm/linux/patch.sh
--- rr264x-linux-src-v1.5.orig/osm/linux/patch.sh	2012-07-19 10:47:35.000000000 +0200
+++ rr264x-linux-src-v1.5/osm/linux/patch.sh	2013-09-29 19:03:26.324066685 +0200
@@ -143,7 +143,7 @@
 		fi
 	fi
 	;;
-	2.6 )
+	2.6 | 3.* )
 	cd ${HPT_ROOT}/lib/linux/free-${ARCH}-regparm0
 	ld -r -o ${ARCH}-${TARGETNAME}.obj ${TARGETMODS}
 	mv ${ARCH}-${TARGETNAME}.obj $KERNELDIR/drivers/scsi/${TARGETNAME}/	
